base_layer/p2p/src/ping_pong.rs

```diff
@@ -0,0 +1,147 @@
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    services::{Service, ServiceContext, ServiceControlMessage},
+    tari_message::{NetMessage, TariMessageType},
+};
+use derive_error::Error;
+use log::*;
+use serde::{export::fmt::Debug, Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use tari_comms::{
+    domain_connector::ConnectorError,
+    message::{Message, MessageFlags, MessageHeader},
+    outbound_message_service::{outbound_message_service::OutboundMessageService, BroadcastStrategy},
+    DomainConnector,
+};
+use tari_utilities::message_format::MessageFormat;
+use threadpool::ThreadPool;
+
+const LOG_TARGET: &'static str = "base_layer::p2p::ping_pong";
+
+#[derive(Serialize, Deserialize)]
+pub enum PingPong {
+    Ping,
+    Pong,
+}
+
+#[derive(Debug, Error)]
+pub enum PingPongError {
+    ConnectorError(ConnectorError),
+}
+
+pub struct PingPongService {
+    interval: Duration,
+}
+
+fn format_err<E: Debug>(err: E) -> String {
+    format!("{:?}", err)
+}
+
+impl PingPongService {
+    pub fn new(interval: Duration) -> Self {
+        Self { interval }
+    }
+
+    fn run(&self, oms: &Arc<OutboundMessageService>, connector: &DomainConnector<'static>) -> Result<(), String> {
+        if let Some((info, msg)) = connector.receive_timeout(self.interval.clone()).map_err(format_err)? {
+            match msg {
+                PingPong::Ping => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Received ping from NodeID {:?}", info.source_identity.node_id
+                    );
+                    // Reply with Pong
+                    let msg = Message::from_message_format(
+                        MessageHeader {
+                            message_type: NetMessage::PingPong,
+                        },
+                        PingPong::Pong,
+                    )
+                    .map_err(format_err)?;
+
+                    oms.send(
+                        BroadcastStrategy::DirectNodeId(info.source_identity.node_id.clone()),
+                        MessageFlags::empty(),
+                        msg.to_binary().map_err(format_err)?,
+                    )
+                    .map_err(format_err)?;
+                },
+                PingPong::Pong => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Received pong from NodeID {:?}", info.source_identity.node_id
+                    );
+                },
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Service for PingPongService {
+    fn get_name(&self) -> String {
+        "ping-pong".to_string()
+    }
+
+    fn get_message_types(&self) -> Vec<TariMessageType> {
+        vec![NetMessage::PingPong.into()]
+    }
+
+    fn execute(&mut self, context: ServiceContext) {
+        let connector = context.create_connector(&NetMessage::PingPong.into()).unwrap();
```

Comment:
Instead of panicking shouldn't this function maybe rather return a result when it break such as on this unwrap

Reply:
Good catch!